### DISCUSSION TITLE
Get a copy of specific ver of a single file from the repository

### DISCUSSION THREAD

**Discussion by chezdrz** [community/community#21700] [New to GitHub]
<p>What is the right command to get a local copy of a specific version of a single file ?</p>
<p>With other revision tools like CVS/SVN, a specific revision number is assigned to a file every time a checkin happens. I don’t see that with Git.</p>
<p>During software development a developer might checkin a file multiple times. But if he/she wants to release a specific ver to Production, what git command do I use to be able to pull out a copy of a specific ver of a file ?</p>
<p>Note:  I do not want to latest ver on the branch or tag, but an intermediate specific ver. </p>

**Comment by seveas** [comment, marked as answer]
<p>Git has various commands to do that. If you want to retrieve that copy without affecting the currently checked out files,</p>
<pre><code>git show
</code></pre>
<p>is the command you’re looking for. For example, to retrieve the file git.c at the revision tagged v2.0.0, you would do:</p>
<pre><code>git show v2.0.0:git.c
</code></pre>
<p>Before the colon, any revision specifier can be used (such as a sha1, a tag or a branch name). After the colon, you put the path to the file you want.</p>

**Comment by chezdrz** [comment]
<p><a class="mention" href="/seveas">@seveas</a> </p>
<p>Thanks for the reply. </p>
<p>I tried show command and it works. However I dont see any revision number listed against each commit as per your example. </p>
<p>I see the SHA value listed against each commit which is a 40 character string, and can get a copy via this </p>
<pre><code>example
git show 85178a160119aa82681e560a63fc217fb7681de9:&lt;file&gt; 
</code></pre>
<p>Is there any setting that I need to turn on which will start labelling revision numbers in place of the SHA value ?</p>

**Comment by seveas** [comment]
<p>Git doesn’t have revision numbers, as they don’t really make sense in Git’s branching and merging model. The sha1’s are unique identifiers for commits, most commands (including git show) will accept an abbreviated version of the sha1 as well for more readable commands. In your example you could use 85178a1 as identifier for instance.</p>

**Comment by The-Futurist** [comment]
<p>This is very easy indeed with SmartGit (which is why I rarely used command line).</p>
<p>Look at this:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/db9042aa1213c7d8c7308ae0a58b93bf2de19f19.png" data-download-href="/uploads/short-url/vklDPRXPi90u9lAgyhAXiVg4aWB.png?dl=1" title="image.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/12262952/181066391-fa287dd1-a70b-4f39-aee0-58661a911e7f.png" alt="image.png" data-base62-sha1="vklDPRXPi90u9lAgyhAXiVg4aWB" width="690" height="433" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/db9042aa1213c7d8c7308ae0a58b93bf2de19f19_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image.png</span><span class="informations">1040×653 61.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>If I click on “Log” (top right) I get this:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/0/0c1cb54fb100df1a4f58bf7080b31078a71fa9d0.png" data-download-href="/uploads/short-url/1J9eGbE9f5Oox6ssaV4EzelQEz6.png?dl=1" title="image.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/12262952/181066395-61165cd1-a76a-4922-ba97-a636ce9d5ba1.png" alt="image.png" data-base62-sha1="1J9eGbE9f5Oox6ssaV4EzelQEz6" width="690" height="447" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/0/0c1cb54fb100df1a4f58bf7080b31078a71fa9d0_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image.png</span><span class="informations">1039×674 83.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Now I can select some arbitrary commit, say 5E74C789, I can then see the files in that commit on the right hand side:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/258b8e0623362286a996c874ee32d0379f78afec.png" data-download-href="/uploads/short-url/5m8EZrFCHVu6ce5Tpen1YS735QE.png?dl=1" title="image.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/12262952/181066396-fe2ed213-38cb-4803-add8-d0976b64de90.png" alt="image.png" data-base62-sha1="5m8EZrFCHVu6ce5Tpen1YS735QE" width="690" height="447" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/258b8e0623362286a996c874ee32d0379f78afec_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image.png</span><span class="informations">1039×674 83.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Now the powerful part, here’s what I get if I right click a file - say SmartPointer.cs:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/9/92748124c29e7538668b09ed1521810a115890ca.png" data-download-href="/uploads/short-url/kTBjRrDeJLJDaotMii9qqiDRsym.png?dl=1" title="image.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/12262952/181066397-cd0df542-12f5-4e0b-a18e-ab0f54c1c19f.png" alt="image.png" data-base62-sha1="kTBjRrDeJLJDaotMii9qqiDRsym" width="690" height="450" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/92748124c29e7538668b09ed1521810a115890ca_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image.png</span><span class="informations">1036×677 83.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You can see “Save As…” if I click that I get this:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/a/a66680b10364f3bc04d42a96fb94b25246fa1d32.png" data-download-href="/uploads/short-url/nK2S0zXnMOhoHDbTroVdB5Ue8ka.png?dl=1" title="image.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/12262952/181066398-7c8e599b-abf9-4d12-ab0e-6237f47d6aac.png" alt="image.png" data-base62-sha1="nK2S0zXnMOhoHDbTroVdB5Ue8ka" width="690" height="449" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/a/a66680b10364f3bc04d42a96fb94b25246fa1d32_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image.png</span><span class="informations">1038×676 84.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>So I can save the file as it looked before that commit or after the changes that were made in that commit.</p>
<p>This is extremely helpful for those ocassions where you simply want to “restore” a single file to some earlier state, you may have made hundreds of changes over many commits but then decide “You know, that constants file with hindsight should have been left alone on this project and the stuff put into a new file” for example.</p>
<p>This is a huge help on those rare ocassions when one wants to do this, in effect this is letting you revert changes to individual files even though those changes might be mixed up with other files.</p>
<p>You do of course get prompted if the file exists when you try to save, letting you verifty that you do indeed want to overwrite the current file, and of course once you do save it you then simply commit that as a change say “Reverted file to its state at XXX”.</p>
